### PR TITLE
SezzleJS 2.0.0

### PR DESCRIPTION
**Pull request by jan-wei**
* Added support for number of payments. Number of payments now should be specified by the `document.sezzleConfig` field `numberOfPayments` (defaults to 4)
* Changed default widget look. Default widget has the current template of `or 4 payments of %%price%% with %%logo%% %%info%%`
* Slightly changed how `renderAwesomeSezzle` works. Previously, the default widget had a widget wrapper wrap two divs by the class `sezzle-button-text` and `sezzle-checkout-button`. By it being two divs, the `sezzle-checkout-button`'s content is being automatically flushed to a newline due to the `display: block` property of divs. This is inconsistent with the default `altVersionTemplate` render as rendering via `altVersionTemplate` just loads the widget as one div containing all the content. Therefore, changes are made to use only the `altVersionTemplate` version for consistency. `altVersionTemplate` is changed to `widgetTemplate` as the rendering of the widget would depend on this template, default or not.
* Added `%%line-break%%` template to the list of assets in `widgetTemplate` as a means of flushing the content to a newline. Previously, `maxWidth` is used to flush the content. This is not ideal as a sweet spot has to be found in order to get the desired widget display. `maxWidth` would also not work on some cases where `%%line-break%%` would always work.
* Changed `Array.from` to `Array.prototype.slice.call / Array.prototype.forEach.call` instead. `Array.prototype.slice.call / Array.prototype.forEach.call` is compatible with ES5 and is much more performant than `Array.from`. Performance test [here](https://jsperf.com/array-from-vs-array-prototype-slice).
* Slightly changed the sharing of work between `init` and `initWidget`. `init` now loads the CSS file first before calling `initWidget` which is the callback. This change is to separate network requests from the actual parsing and rendering of the widget. This enables the developer to just call `initWidget` for any re-renders as all assets have already been loaded on the current session. This also increases performance as network latency is not present in re-renders.
* Changed `hideClasses` to `hideElements`. `hideElements` are targeted via our version of xPath instead of just a class for better flexibility. Also changed the targeting of `ignoredPriceElements` to use the `getAllPriceElements` function instead of building the query string and calling via `querySelectorAll` as that is incorrect (`nth-of-type()` or `nth-child()` CSS selector does not work according to the way `getAllPriceElements` targets via our version of xPath). As a consequence, the `getAllPriceElements` function is renamed to a more general `getElementsByXPath` as it is being used for various functions, not only retrieving the price element.
* Defaulted the widget alignment to `auto`. An auto alignment would align the widget according to the computed `textAlign` property of the targeted price element. This addition would effectively cease the need of `alignmentSwitchMinWidth` and `alignmentSwitchType` as alignment switches do not need to be accounted for already. This should increase efficiency and productivity. Maintained the old options (`center`, `left`, `right`) and also the alignment switch features in case the widget does not align properly via `auto`.
* Removed the `of ` string from the `%%price%%` template. This would unbold the `of ` string, which is intended.
* Slightly changed CSS styling. Replaced `fontSize` with `widgetSize`. `fontSize` has been replaced because other assets in the widget such as the logo does not scale with increasing / decreasing font size. CSS scale transforms are instead use to scale the widget size. The transform origin would be equal to `top + this.widgetAlignment`.
* Deprecated `widthType` and `hidePrice` as both are redundant and not in use anymore.

### CODE REVIEW

**Comment by jan-wei** [comment]
**Test Cases**

All the test case infos will be detailed here including :
* What the test case aims to test for.
* Test input
* Expected output
* Status of test

All cases will be tested on Chrome 69.0.3497.100, Firefox 61.0.2, IE11, Safari 11.1.2 and Edge 17. It should be understood that if it is stated that the test passed, it passed on all these versions.

**Comment by jan-wei** [comment]
### Test Case 1

#### Purpose
Test multiple `targetXPath`, `renderToPath` to access recent changes from `Array.from` to `Array.prototype.slice.call / Array.prototype.forEach.call`

#### Test Location
https://streetsole.com/collections/on-sale/products/pacific-tour-zip-hoodie
#### Test Input
```javascript
document.sezzleConfig =
{
     targetXPath: ["#ProductPrice-product-template", '#ComparePrice-product-template/.money'],
     renderToPath: [".", "../../.."],
}
```

#### Expected Output
Two widgets rendered, one for the sale price and one for the compared price. The widget for the compared price should be rendered far down the page as the path corresponding to it.

#### Status
Passed

**Comment by rishimukherjee** [comment]
@jan-wei Removing `hideClasses` and changing it to `hideElements` is not backward compatible right?

**Comment by rishimukherjee** [comment]
Same question about `widgetSize` and `fontSize`. Because with this, we have to change configurations of merchants isn't it? @jan-wei

**Comment by jan-wei** [comment]
> @jan-wei Removing `hideClasses` and changing it to `hideElements` is not backward compatible right?

nope unfortunately :( These changes are just meant for current merchants moving forward, same with `widgetSize`. i hope that is okay. Unless we reintegrate for everyone, which I dont think we will :)

File: sezzle.js [hunk 1]
```diff
@@ -1,282 +1,229 @@
...
-  this.rendertopath = [];
-  if (options.renderToPath) {
-    if (typeof(options.renderToPath) === 'string') {
-      // Only one respective render location is given
-      this.rendertopath.push(options.renderToPath);
-    } else {
-      // options.renderToPath is an array of x-paths
-      this.rendertopath = options.renderToPath;
+    if(this.splitPriceElementsOn) {
+        this.widgetTemplate = this.widgetTemplate.map(function(subtemplate, index) {
```
Comment (sanbornsen) [inline sezzle.js]: There is no use of `index` in the function.

File: sezzle.js [hunk 2]
```diff
@@ -1,282 +1,229 @@
...
-    if (index in this.rendertopath) {
-      this.rendertopath[index] =
-        this.rendertopath[index].trim() != '' ?
-          this.rendertopath[index].trim() : null;
-    } else {
-      this.rendertopath.push(null);
+
+    // Search for price elements. If found, assume there is only one in this page
+    this.hasPriceClassElement = false;
+    this.priceElements = Array.prototype.slice.call(document.getElementsByClassName(this.priceElementClass));
```
Comment (sanbornsen) [inline sezzle.js]: Line break needed

File: sezzle.js [hunk 3]
```diff
@@ -1,282 +1,229 @@
...
+    this.renderElements = Array.prototype.slice.call(document.getElementsByClassName(this.sezzleWidgetContainerClass));
```
Comment (sanbornsen) [inline sezzle.js]: Line break needed

File: sezzle.js (hunk 1 above)
Comment (MartaJopp) [reply sezzle.js]: removed

File: sezzle.js [hunk 4]
```diff
@@ -1,282 +1,229 @@
...
+            // Only one x-path is given
+            this.ignoredPriceElements.push(options.ignoredPriceElements.split('/').filter(function(subpath) {
+                 return subpath !== "";
+            }));
+        }
+        else {
+            // options.ignoredPriceElements is an array of x-paths
+            this.ignoredPriceElements = options.ignoredPriceElements.map(function(path){
+                return path.split('/');
+            }).filter(function(subpath) {
```
[the remark this answers is not in the document]
Comment (MartaJopp) [reply sezzle.js]: updated

File: sezzle.js [hunk 5]
```diff
@@ -1,282 +1,229 @@
...
+            // Only one x-path is given
+            this.hideElements.push(options.hideClasses.split('/').filter(function(subpath) {
+                 return subpath !== "";
+            }));
+        }
+        else {
+            // options.hideClasses is an array of x-paths
+            this.hideElements = options.hideClasses.map(function(path){
+                return path.split('/');
+            }).filter(function(subpath) {
```
[the remark this answers is not in the document]
Comment (MartaJopp) [reply sezzle.js]: updated

File: sezzle.js [hunk 6]
```diff
@@ -530,204 +468,142 @@ SezzleJS.prototype.renderAwesomeSezzle = function(element, renderelement, index
...
+  //get the alignment of the widget (if widgetAlignment is auto)
+  // the alignment, when set to auto follows the text-align property of the price element
+  if (this.alignment === 'auto') {
+    this.alignment = this.guessWidgetAlignment(element)
+  }
+
   // root node for sezzle
   var sezzle = document.createElement('div');
-  sezzle.className = "sezzle-shopify-info-button"
+  sezzle.className = 'sezzle-shopify-info-button'
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) missing in this line

File: sezzle.js [hunk 7]
```diff
@@ -530,204 +468,142 @@ SezzleJS.prototype.renderAwesomeSezzle = function(element, renderelement, index
   // Get element to be rendered with sezzle's widget
   var parent = renderelement;
 
+  //get the alignment of the widget (if widgetAlignment is auto)
+  // the alignment, when set to auto follows the text-align property of the price element
+  if (this.alignment === 'auto') {
+    this.alignment = this.guessWidgetAlignment(element)
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) missing in this line

File: sezzle.js [hunk 8]
```diff
@@ -530,204 +468,142 @@ SezzleJS.prototype.renderAwesomeSezzle = function(element, renderelement, index
...
+  sezzleButtonText.className = 'sezzle-button-text';
+  this.addCSSCustomisation(sezzleButtonText);
+
+  this.widgetTemplate.forEach(function (subtemplate) {
+    switch (subtemplate) {
+      case 'price':
+        var priceSpanNode = document.createElement('span');
+        priceSpanNode.className = 'sezzle-payment-amount sezzle-button-text sezzleindex-' + index;
+        var priceValueText = document.createTextNode(this.getFormattedPrice(element));
+        priceSpanNode.appendChild(priceValueText)
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) missing in this line

File: sezzle.js [hunk 9]
```diff
@@ -530,204 +468,142 @@ SezzleJS.prototype.renderAwesomeSezzle = function(element, renderelement, index
...
+        sezzleButtonText.appendChild(learnMoreNode);
+        break;
 
-    // Loge node first child level - 1.1.4
-    var logoNode4 = document.createElement("div");
-    logoNode4.className = "sezzle-know-more";
+      case 'info':
+        var infoIconNode = document.createElement('code');
+        infoIconNode.className = 'sezzle-modal-link sezzle-info-icon';
+        infoIconNode.innerHTML = '&#9432;'
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) missing in this line

File: sezzle.js [hunk 10]
```diff
@@ -530,204 +468,142 @@ SezzleJS.prototype.renderAwesomeSezzle = function(element, renderelement, index
...
+        sezzleButtonText.appendChild(infoIconNode);
+        break;
 
-    // Logo node first child text - 1.1.4.1
-    var logoNode4Text = document.createTextNode(' Learn more');
-    logoNode4.appendChild(logoNode4Text); // 1.1.4
+      case 'question-mark':
+        var questionMarkIconNode = document.createElement('img')
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) missing in this line

File: sezzle.js [hunk 11]
```diff
@@ -530,204 +468,142 @@ SezzleJS.prototype.renderAwesomeSezzle = function(element, renderelement, index
...
+        questionMarkIconNode.className = 'sezzle-modal-link sezzle-question-mark-icon';
+        questionMarkIconNode.src = 'https://d2uyik3j5wol98.cloudfront.net/images/question_mark_black.png'
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) missing in this line

File: sezzle.js [hunk 12]
```diff
@@ -530,204 +468,142 @@ SezzleJS.prototype.renderAwesomeSezzle = function(element, renderelement, index
...
+        sezzleButtonText.appendChild(questionMarkIconNode);
+        break;
 
-    // Add logeNode1 to logoNode level - 1.1
-    logoNode.appendChild(logoNode4);
+      case 'afterpay-logo':
+        var apNode = document.createElement('img');
+        apNode.className = 'sezzle-afterpay-logo'
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) missing in this line

File: sezzle.js [hunk 13]
```diff
@@ -530,204 +468,142 @@ SezzleJS.prototype.renderAwesomeSezzle = function(element, renderelement, index
...
+        apNode.src = 'https://d34uoa9py2cgca.cloudfront.net/sezzle-credit-website-assets/ap-logo-widget.png';
+        sezzleButtonText.appendChild(apNode);
+        break;
 
-    // Adding logoNode to main node
-    node.appendChild(logoNode);
+      case 'afterpay-logo-grey':
+        var apNode = document.createElement('img');
+        apNode.className = 'sezzle-afterpay-logo'
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) missing in this line

File: sezzle.js [hunk 14]
```diff
@@ -530,204 +468,142 @@ SezzleJS.prototype.renderAwesomeSezzle = function(element, renderelement, index
...
-              }
-              else {
-                priceSplitText = priceSplitText + ' ' + this.splitPriceElementsOn + ' ' + this.getFormattedPrice(elem);
-              }
-            }.bind(this))
-          }
+      case 'afterpay-info-icon':
+        var apInfoIconNode = document.createElement('code');
+        apInfoIconNode.className = 'ap-modal-info-link';
+        apInfoIconNode.innerHTML = '&#9432;'
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) missing in this line

File: sezzle.js [hunk 15]
```diff
@@ -530,204 +468,142 @@ SezzleJS.prototype.renderAwesomeSezzle = function(element, renderelement, index
...
+        sezzleButtonText.appendChild(apInfoIconNode);
+        break;
 
-          var priceSplitTextNode = document.createTextNode(priceSplitText);
-          priceSplitNode.appendChild(priceSplitTextNode);
+      case 'price-split':
+        var priceSplitNode = document.createElement('span');
+        priceSplitNode.className = 'sezzle-payment-amount sezzle-price-split sezzleindex-' + index;
+        var priceElemTexts = element.innerText.split(this.splitPriceElementsOn);
+        var priceSplitText = ''
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) missing in this line

File: sezzle.js [hunk 16]
```diff
@@ -530,204 +468,142 @@ SezzleJS.prototype.renderAwesomeSezzle = function(element, renderelement, index
...
+        if (priceElemTexts.length == 1) { //if the text is not being splitted (this check is needed in order to support sites with multiple types of product pricing)
+          //give the original element in the case there might be some ignored elements present
+          priceSplitText = this.getFormattedPrice(element)
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) missing in this line

File: sezzle.js [hunk 17]
```diff
@@ -530,204 +468,142 @@ SezzleJS.prototype.renderAwesomeSezzle = function(element, renderelement, index
...
+            priceElemSpan.innerText = text;
+            priceElems.push(priceElemSpan);
+          });
+          priceElems.forEach(function (elem, index) {
+            if (index == 0) {
+              priceSplitText = this.getFormattedPrice(elem);
+            } else {
+              priceSplitText = priceSplitText + ' ' + this.splitPriceElementsOn + ' ' + this.getFormattedPrice(elem);
+            }
+          }.bind(this))
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) missing in this line

File: sezzle.js [hunk 18]
```diff
@@ -777,11 +656,11 @@ SezzleJS.prototype.getElementToRender = function(element, index) {
         toRenderElement =
           toRenderElement.children.length > 0 ?
             toRenderElement.firstElementChild :
-            null ;
+            null;
         this.widgetIsFirstChild = true
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) missing in this line

File: sezzle.js [hunk 19]
```diff
@@ -830,122 +708,89 @@ SezzleJS.prototype.insertAsFirstChild = function(element, referenceElement) {
...
-  var numberOfPayments = 4
-  if (document.sezzleConfig && document.sezzleConfig.numberOfPayments) {
-      numberOfPayments = document.sezzleConfig.numberOfPayments
-  }
+  // array of strings that come up inside of elements that we want to make sure to strip out
+  var ignoredPriceStrings = [
+    'Subtotal',
+    'Total:',
+    'Sold Out',
+  ]
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) missing in this line

File: sezzle.js [hunk 20]
```diff
@@ -954,131 +799,90 @@ SezzleJS.prototype.getFormattedPrice = function(element) {
...
-SezzleJS.prototype.deleteObserver = new MutationObserver(function(mutations) {
+SezzleJS.prototype.deleteObserver = new MutationObserver(function (mutations) {
   // Get the mutations which have both added and removed nodes
-  var removedAddedMutations = mutations
-    .filter(function(mutation) {
-      return mutation.removedNodes.length && mutation.addedNodes.length
-    });
-  // if there is at least on removed added mutation
+  var removedAddedMutations = mutations.filter(function (mutation) {
+    return mutation.removedNodes.length && mutation.addedNodes.length
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) missing in this line

File: sezzle.js [hunk 21]
```diff
@@ -1105,75 +909,123 @@ SezzleJS.prototype.startObserve = function(element) {
...
-      modalNode.getElementsByClassName('sezzle-checkout-modal')[0].className = "sezzle-checkout-modal sezzle-checkout-modal-hidden";
+      modalNode.getElementsByClassName('sezzle-checkout-modal')[0].className = 'sezzle-checkout-modal sezzle-checkout-modal-hidden';
     });
+  });
+
+  // Event listener to prevent close in modal if click happens within sezzle-checkout-modal
+  document.getElementsByClassName('sezzle-checkout-modal')[0].addEventListener('click', function (event) {
+    // stop propagating the event to the parent sezzle-checkout-modal-lightbox to prevent the closure of the modal
+    event.stopPropagation();
+  })
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) is missing in this line

File: sezzle.js [hunk 22]
```diff
@@ -1105,75 +909,123 @@ SezzleJS.prototype.startObserve = function(element) {
...
...: auto;"><div style="display: flex; place-content: center; align-items: center; width: 100%; min-height: 100%; background-color: rgba(0, 0, 0, 0.8);"><div style="position: relative; background-color: rgb(255, 255, 255);"><a href="' + this.afterpayModalClickURL + '" target="_blank" style="display: block;"><img src="//cdn.shopify.com/s/files/1/0105/4102/t/53/assets/PX_S18AfterPay-Horz.jpg" style="display: block; width: 100%;"></a><a href="#" style="position: absolute; right: 8px; top: 8px;"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 32 32" version="1.1" width="32px" height="32px"><g id="surface1"><path style=" " d="M 16 3 C 8.832031 3 3 8.832031 3 16 C 3 23.167969 8.832031 29 16 29 C 23.167969 29 29 23.167969 29 16 C 29 8.832031 23.167969 3 16 3 Z M 16 5 C 22.085938 5 27 9.914063 27 16 C 27 22.085938 22.085938 27 16 27 C 9.914063 27 5 22.085938 5 16 C 5 9.914063 9.914063 5 16 5 Z M 12.21875 10.78125 L 10.78125 12.21875 L 14.5625 16 L 10.78125 19.78125 L 12.21875 21.21875 L 16 17.4375 L 19.78125 21.21875 L 21.21875 19.78125 L 17.4375 16 L 21.21875 12.21875 L 19.78125 10.78125 L 16 14.5625 Z "></path></g></svg></a></div></div></div>'
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) is missing in this line

File: sezzle.js [hunk 23]
```diff
@@ -1271,54 +1116,50 @@ SezzleJS.prototype.replaceBanner = function() {
...
 */
-SezzleJS.prototype.logEvent = function(eventName) {
-
-  if(this.fingerprint == null){
-    this.getFingerprint(function(fingerprint) {
-
+SezzleJS.prototype.logEvent = function (eventName) {
+  if (this.fingerprint == null) {
+    this.getFingerprint(function (fingerprint) {
       this.fingerprint = fingerprint
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) is missing in this line

File: sezzle.js [hunk 24]
```diff
@@ -1331,27 +1172,26 @@ SezzleJS.prototype.postEvent = function(payload) {
 /*
 * Get Fingerprint
 */
-SezzleJS.prototype.getFingerprint = function(callback) {
-  if(typeof window.Fingerprint2 === 'undefined' || window.Fingerprint2.VERSION != "1.4.1") {
-    var script = document.createElement("script")
-    script.type = "text/javascript";
+SezzleJS.prototype.getFingerprint = function (callback) {
+  if (typeof window.Fingerprint2 === 'undefined' || window.Fingerprint2.VERSION != '1.4.1') {
+    var script = document.createElement('script')
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) is missing in this line

File: sezzle.js [hunk 25]
```diff
@@ -1360,56 +1200,54 @@ SezzleJS.prototype.getFingerprint = function(callback) {
...
...d|un)|haie|hcit|hd\-(m|p|t)|hei\-|hi(pt|ta)|hp( i|ip)|hs\-c|ht(c(\-| |_|a|g|p|s|t)|tp)|hu(aw|tc)|i\-(20|go|ma)|i230|iac( |\-|\/)|ibro|idea|ig01|ikom|im1k|inno|ipaq|iris|ja(t|v)a|jbro|jemu|jigs|kddi|keji|kgt( |\/)|klon|kpt |kwc\-|kyo(c|k)|le(no|xi)|lg( g|\/(k|l|u)|50|54|\-[a-w])|libw|lynx|m1\-w|m3ga|m50\/|ma(te|ui|xo)|mc(01|21|ca)|m\-cr|me(rc|ri)|mi(o8|oa|ts)|mmef|mo(01|02|bi|de|do|t(\-| |o|v)|zz)|mt(50|p1|v )|mwbp|mywa|n10[0-2]|n20[2-3]|n30(0|2)|n50(0|2|5)|n7(0(0|1)|10)|ne((c|m)\-|on|tf|wf|wg|wt)|nok(6|i)|nzph|o2im|op(ti|wv)|oran|owg1|p800|pan(a|d|t)|pdxg|pg(13|\-([1-8]|c))|phil|pire|pl(ay|uc)|pn\-2|po(ck|rt|se)|prox|psio|pt\-g|qa\-a|qc(07|12|21|32|60|\-[2-7]|i\-)|qtek|r380|r600|raks|rim9|ro(ve|zo)|s55\/|sa(ge|ma|mm|ms|ny|va)|sc(01|h\-|oo|p\-)|sdk\/|se(c(\-|0|1)|47|mc|nd|ri)|sgh\-|shar|sie(\-|m)|sk\-0|sl(45|id)|sm(al|ar|b3|it|t5)|so(ft|ny)|sp(01|h\-|v\-|v )|sy(01|mb)|t2(18|50)|t6(00|10|18)|ta(gt|lk)|tcl\-|tdg\-|tel(i|m)|tim\-|t\-mo|to(pl|sh)|ts(70|m\-|m3|m5)|tx\-9|up(\.b|g1|si)|utst|v400|v750|veri|vi(rg|te)|vk(40|5[0-3]|\-v)|vm40|voda|vulc|vx(52|53|60|61|70|80|81|83|85|98)|w3c(\-| )|webc|whit|wi(g |nc|nw)|wmlb|wonu|x700|yas\-|your|zeto|zte\-/i.test(navigator.userAgent.substr(0, 4))
```
Comment (sanbornsen) [inline sezzle.js]: A semicolon (`;`) is missing in this line

**Pull request merged**